Compiling anything that imports `wasi_experimental_http_wasmtime` fails on Windows

This is because the `wasi_experimental_http_wasmtime` crate imports two functions from the `wasi_experimental_http` crate, which in turn relies on external symbols (the host functions related to HTTP).

And even if the two functions imported don't actually call any of the external symbols, the Windows linker still fails.

I will investigate if this is a known issue with the linker on Windows, but removing the dependency to `wasi_experimental_http` solves the issue, and the two functions are simple enough that we can just copy-paste.

```
  = note: libwasi_experimental_http-c7b972a7a58f275d.rlib(wasi_experimental_http-c7b972a7a58f275d.39shyjq0d6lqu8qa.rcgu.o) : 
error LNK2019: unresolved external symbol __imp_req referenced in function _ZN22wasi_experimental_http3raw3req17h295ccbc09f2fe42dE
          libwasi_experimental_http-c7b972a7a58f275d.rlib(wasi_experimental_http-c7b972a7a58f275d.39shyjq0d6lqu8qa.rcgu.o) : 
error LNK2019: unresolved external symbol __imp_header_get referenced in function _ZN22wasi_experimental_http3raw10header_get17h1870664f7a8646ffE
          libwasi_experimental_http-c7b972a7a58f275d.rlib(wasi_experimental_http-c7b972a7a58f275d.39shyjq0d6lqu8qa.rcgu.o) : 
error LNK2019: unresolved external symbol __imp_headers_get_all referenced in function _ZN22wasi_experimental_http3raw15headers_get_all17hb2f4ccc303974529E
          libwasi_experimental_http-c7b972a7a58f275d.rlib(wasi_experimental_http-c7b972a7a58f275d.39shyjq0d6lqu8qa.rcgu.o) : 
error LNK2019: unresolved external symbol __imp_body_read referenced in function _ZN22wasi_experimental_http3raw9body_read17h4cbd4ef921d4891cE
          D:\projects\deislabs\wasi-experimental-http\target\debug\deps\integration-159d744b0b153291.exe : fatal error LNK1120: 4 unresolved externals
```

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -12,6 +12,7 @@ env:
 jobs:
   build:
     strategy:
+      fail-fast: false
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
     runs-on: ${{ matrix.os }}
@@ -21,4 +22,15 @@ jobs:
     - name: Build
       run: rustup target add wasm32-wasi && cargo build --verbose
     - name: Run simple test
-      run: cargo test --all -- --nocapture
+      # run: cargo test -- --nocapture
+      run:   |
+            if [ "$RUNNER_OS" == "Windows" ]; then
+                  # Attempting to run the unit tests on Windows
+                  # results in the linker failing because it cannot
+                  # compile the Rust client library because of the
+                  # missing external symbols, so we are only running
+                  # the integration tests on Windows.
+                  cargo test -- --nocapture
+            else
+                  cargo test --all -- --nocapture1
+            fi

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -11,12 +11,14 @@ env:
 
 jobs:
   build:
-
-    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+    runs-on: ${{ matrix.os }}
 
     steps:
     - uses: actions/checkout@v2
     - name: Build
-      run: rustup target add wasm32-wasi && cargo install witx-codegen && cargo build --verbose
+      run: rustup target add wasm32-wasi && cargo build --verbose
     - name: Run simple test
       run: cargo test --all -- --nocapture

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -22,7 +22,7 @@ jobs:
     - name: Build
       run: rustup target add wasm32-wasi && cargo build --verbose
     - name: Run simple test
-      # run: cargo test -- --nocapture
+      shell: bash
       run:   |
             if [ "$RUNNER_OS" == "Windows" ]; then
                   # Attempting to run the unit tests on Windows
@@ -32,5 +32,5 @@ jobs:
                   # the integration tests on Windows.
                   cargo test -- --nocapture
             else
-                  cargo test --all -- --nocapture1
+                  cargo test --all -- --nocapture
             fi

--- a/.gitignore
+++ b/.gitignore
@@ -5,3 +5,5 @@ build/
 node_modules/
 package-lock.json
 build
+*.exe
+*.pdb

--- a/build.rs
+++ b/build.rs
@@ -1,4 +1,4 @@
-use std::process;
+use std::process::{self, Command};
 
 const TESTS_DIR: &str = "tests";
 const RUST_EXAMPLE: &str = "rust";
@@ -8,6 +8,8 @@ const RUST_GUEST_RAW: &str = "crates/wasi-experimental-http/src/raw.rs";
 const AS_GUEST_RAW: &str = "crates/as/raw.ts";
 const MD_GUEST_API: &str = "witx/readme.md";
 
+const WITX_CODEGEN_VERSION: &str = "0.10.4";
+
 fn main() {
     println!("cargo:rerun-if-changed=build.rs");
     println!("cargo:rerun-if-changed=tests/rust/src/lib.rs");
@@ -16,44 +18,28 @@ fn main() {
     println!("cargo:rerun-if-changed=crates/as/index.ts");
     println!("cargo:rerun-if-changed=witx/wasi_experimental_http.witx");
 
-    generate_from_witx("rust".to_string(), RUST_GUEST_RAW.to_string());
-    generate_from_witx("assemblyscript".to_string(), AS_GUEST_RAW.to_string());
-    generate_from_witx("markdown".to_string(), MD_GUEST_API.to_string());
+    generate_from_witx("rust", RUST_GUEST_RAW);
+    generate_from_witx("assemblyscript", AS_GUEST_RAW);
+    generate_from_witx("markdown", MD_GUEST_API);
 
-    cargo_build_example(TESTS_DIR.to_string(), RUST_EXAMPLE.to_string());
-    as_build_example(TESTS_DIR.to_string(), AS_EXAMPLE.to_string());
+    cargo_build_example(TESTS_DIR, RUST_EXAMPLE);
+    as_build_example(TESTS_DIR, AS_EXAMPLE);
 }
 
-fn cargo_build_example(dir: String, example: String) {
+fn cargo_build_example(dir: &str, example: &str) {
     let dir = format!("{}/{}", dir, example);
 
-    let mut cmd = process::Command::new("cargo");
-    cmd.current_dir(dir);
-    cmd.stdout(process::Stdio::piped());
-    cmd.stderr(process::Stdio::piped());
-    cmd.arg("build")
-        .arg("--target")
-        .arg("wasm32-wasi")
-        .arg("--release");
-    cmd.output().unwrap();
+    run(
+        vec!["cargo", "build", "--target", "wasm32-wasi", "--release"],
+        Some(dir),
+    );
 }
 
-fn as_build_example(dir: String, example: String) {
+fn as_build_example(dir: &str, example: &str) {
     let dir = format!("{}/{}", dir, example);
 
-    let mut cmd = process::Command::new("npm");
-    cmd.current_dir(dir.clone());
-    cmd.stdout(process::Stdio::piped());
-    cmd.stderr(process::Stdio::piped());
-    cmd.arg("install");
-    cmd.output().unwrap();
-
-    let mut cmd = process::Command::new("npm");
-    cmd.current_dir(dir);
-    cmd.stdout(process::Stdio::piped());
-    cmd.stderr(process::Stdio::piped());
-    cmd.arg("run").arg("asbuild");
-    cmd.output().unwrap();
+    run(vec!["npm", "install"], Some(dir.clone()));
+    run(vec!["npm", "run", "asbuild"], Some(dir));
 }
 
 fn check_witx_codegen() {
@@ -61,31 +47,64 @@ fn check_witx_codegen() {
         Ok(_) => {
             eprintln!("witx-codegen already installed");
         }
-        Err(e) => {
-            if let std::io::ErrorKind::NotFound = e.kind() {
-                let mut cmd = process::Command::new("cargo");
-                cmd.stdout(process::Stdio::piped());
-                cmd.stderr(process::Stdio::piped());
-                cmd.arg("install").arg("witx-codegen");
-                cmd.output().unwrap();
-            } else {
-                eprintln!("cannot find or install witx-codegen: {}", e);
-            }
+        Err(_) => {
+            println!("cannot find witx-codegen, attempting to install");
+            run(
+                vec![
+                    "cargo",
+                    "install",
+                    "witx-codegen",
+                    "--version",
+                    WITX_CODEGEN_VERSION,
+                ],
+                None,
+            );
         }
     }
 }
 
-fn generate_from_witx(codegen_type: String, output: String) {
+fn generate_from_witx(codegen_type: &str, output: &str) {
     check_witx_codegen();
-    let mut cmd = process::Command::new("witx-codegen");
+
+    run(
+        vec![
+            "witx-codegen",
+            "--output-type",
+            codegen_type,
+            "--output",
+            output,
+            "witx/wasi_experimental_http.witx",
+        ],
+        None,
+    );
+}
+
+fn run<S: Into<String> + std::convert::AsRef<std::ffi::OsStr>>(args: Vec<S>, dir: Option<String>) {
+    let mut cmd = Command::new(get_os_process());
     cmd.stdout(process::Stdio::piped());
     cmd.stderr(process::Stdio::piped());
-    cmd.arg("--output-type")
-        .arg(codegen_type)
-        .arg("--output")
-        .arg(output)
-        .arg("witx/wasi_experimental_http.witx");
 
-    eprintln!("test {:#?}", cmd);
+    if let Some(dir) = dir {
+        cmd.current_dir(dir);
+    };
+
+    cmd.arg("-c");
+    cmd.arg(
+        args.into_iter()
+            .map(Into::into)
+            .collect::<Vec<String>>()
+            .join(" "),
+    );
+
+    println!("running {:#?}", cmd);
+
     cmd.output().unwrap();
+}
+
+fn get_os_process() -> String {
+    if cfg!(target_os = "windows") {
+        String::from("powershell.exe")
+    } else {
+        String::from("/bin/bash")
+    }
 }

--- a/build.rs
+++ b/build.rs
@@ -79,7 +79,7 @@ fn generate_from_witx(codegen_type: &str, output: &str) {
     );
 }
 
-fn run<S: Into<String> + std::convert::AsRef<std::ffi::OsStr>>(args: Vec<S>, dir: Option<String>) {
+fn run<S: Into<String> + AsRef<std::ffi::OsStr>>(args: Vec<S>, dir: Option<String>) {
     let mut cmd = Command::new(get_os_process());
     cmd.stdout(process::Stdio::piped());
     cmd.stderr(process::Stdio::piped());

--- a/crates/wasi-experimental-http-wasmtime/Cargo.toml
+++ b/crates/wasi-experimental-http-wasmtime/Cargo.toml
@@ -20,5 +20,4 @@ url = "2.2.1"
 wasmtime = "0.26"
 wasmtime-wasi = "0.26"
 wasi-common = "0.26"
-wasi-experimental-http = { version = "0.2", path = "../wasi-experimental-http" }
 tracing = { version = "0.1", features = ["log"] }

--- a/crates/wasi-experimental-http-wasmtime/src/lib.rs
+++ b/crates/wasi-experimental-http-wasmtime/src/lib.rs
@@ -6,7 +6,6 @@ use reqwest::{Client, Method};
 use std::{cell::RefCell, collections::HashMap, rc::Rc, str::FromStr};
 use tokio::runtime::Handle;
 use url::Url;
-// use wasi_experimental_http::header_map_to_string;
 use wasmtime::*;
 
 const MEMORY: &str = "memory";
@@ -554,7 +553,7 @@ fn is_allowed(url: &str, allowed_hosts: Option<&[String]>) -> Result<bool, HttpE
 // https://github.com/rust-lang/rust/issues/86125
 
 /// Decode a header map from a string.
-pub fn string_to_header_map(s: &str) -> Result<HeaderMap, Error> {
+fn string_to_header_map(s: &str) -> Result<HeaderMap, Error> {
     let mut headers = HeaderMap::new();
     for entry in s.lines() {
         let mut parts = entry.splitn(2, ':');
@@ -570,7 +569,7 @@ pub fn string_to_header_map(s: &str) -> Result<HeaderMap, Error> {
 }
 
 /// Encode a header map as a string.
-pub fn header_map_to_string(hm: &HeaderMap) -> Result<String, Error> {
+fn header_map_to_string(hm: &HeaderMap) -> Result<String, Error> {
     let mut res = String::new();
     for (name, value) in hm
         .iter()

--- a/crates/wasi-experimental-http-wasmtime/src/lib.rs
+++ b/crates/wasi-experimental-http-wasmtime/src/lib.rs
@@ -1,12 +1,12 @@
 use anyhow::Error;
 use bytes::Bytes;
 use futures::executor::block_on;
-use http::{HeaderMap, HeaderValue};
+use http::{header::HeaderName, HeaderMap, HeaderValue};
 use reqwest::{Client, Method};
 use std::{cell::RefCell, collections::HashMap, rc::Rc, str::FromStr};
 use tokio::runtime::Handle;
 use url::Url;
-use wasi_experimental_http::header_map_to_string;
+// use wasi_experimental_http::header_map_to_string;
 use wasmtime::*;
 
 const MEMORY: &str = "memory";
@@ -233,7 +233,7 @@ impl HostCalls {
         let method = Method::from_str(string_from_memory(&memory, method_ptr, method_len)?)
             .map_err(|_| HttpError::InvalidMethod)?;
         let req_body = slice_from_memory(&memory, req_body_ptr, req_body_len)?;
-        let headers = wasi_experimental_http::string_to_header_map(string_from_memory(
+        let headers = string_to_header_map(string_from_memory(
             &memory,
             req_headers_ptr,
             req_headers_len,
@@ -544,6 +544,52 @@ fn is_allowed(url: &str, allowed_hosts: Option<&[String]>) -> Result<bool, HttpE
         }
         None => Ok(false),
     }
+}
+
+// The following two functions are copied from the `wasi_experimental_http`
+// crate, because the Windows linker apparently cannot handle unresolved
+// symbols from a crate, even when the caller does not actually use any of the
+// external symbols.
+//
+// https://github.com/rust-lang/rust/issues/86125
+
+/// Decode a header map from a string.
+pub fn string_to_header_map(s: &str) -> Result<HeaderMap, Error> {
+    let mut headers = HeaderMap::new();
+    for entry in s.lines() {
+        let mut parts = entry.splitn(2, ':');
+        #[allow(clippy::clippy::clippy::or_fun_call)]
+        let k = parts.next().ok_or(anyhow::format_err!(
+            "Invalid serialized header: [{}]",
+            entry
+        ))?;
+        let v = parts.next().unwrap();
+        headers.insert(HeaderName::from_str(k)?, HeaderValue::from_str(v)?);
+    }
+    Ok(headers)
+}
+
+/// Encode a header map as a string.
+pub fn header_map_to_string(hm: &HeaderMap) -> Result<String, Error> {
+    let mut res = String::new();
+    for (name, value) in hm
+        .iter()
+        .map(|(name, value)| (name.as_str(), std::str::from_utf8(value.as_bytes())))
+    {
+        let value = value?;
+        anyhow::ensure!(
+            !name
+                .chars()
+                .any(|x| x.is_control() || "(),/:;<=>?@[\\]{}".contains(x)),
+            "Invalid header name"
+        );
+        anyhow::ensure!(
+            !value.chars().any(|x| x.is_control()),
+            "Invalid header value"
+        );
+        res.push_str(&format!("{}:{}\n", name, value));
+    }
+    Ok(res)
 }
 
 #[test]

--- a/tests/as/index.ts
+++ b/tests/as/index.ts
@@ -21,8 +21,10 @@ export function get(): void {
     .send();
 
   check(res, 200, "content-type");
-  let body = res.bodyReadAll();
-  if (String.UTF8.decode(body.buffer) != '"OK"') {
+  let bytes = res.bodyReadAll();
+  let body = String.UTF8.decode(bytes.buffer);
+  if (!body.includes("OK")) {
+    Console.write("got " + body);
     abort();
   }
   res.close();
@@ -61,7 +63,7 @@ function check(
   }
 
   let headers = res.headerGetAll();
-  if(headers.size == 0) {
+  if (headers.size == 0) {
     abort();
   }
 }

